<a href="https://colab.research.google.com/github/yli1048/yli1048/blob/602/07_assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Assignment 7**

# **Weeks 8 & 9 - Pandas**
* In this homework assignment, you will explore and analyze a public dataset of your choosing. Since this assignment is “open-ended” in nature, you are free to expand upon the requirements below. However, you must meet the minimum requirments as indicated in each section.

* You must use Pandas as the **primary tool** to process your data.

* The preferred method for this analysis is in a .ipynb file. Feel free to use whichever platform of your choosing.  
 * https://www.youtube.com/watch?v=inN8seMm7UI (Getting started with Colab).

* Your data should need some "work", or be considered "dirty".  You must show your skills in data cleaning/wrangling.

### **Some data examples:**
•	https://www.data.gov/

•	https://opendata.cityofnewyork.us/

•	https://datasetsearch.research.google.com/

•	https://archive.ics.uci.edu/ml/index.php

### **Resources:**

•	https://pandas.pydata.org/pandas-docs/stable/getting_started/10min.html

•	https://pandas.pydata.org/pandas-docs/stable/user_guide/visualization.html


### **Headings or comments**
**You are required to make use of comments, or headings for each section.  You must explain what your code is doing, and the results of running your code.**  Act as if you were giving this assignment to your manager - you must include clear and descriptive information for each section.

### **You may work as a group or indivdually on this assignment.**


# Introduction

In this section, please describe the dataset you are using.  Include a link to the source of this data.  You should also provide some explanation on why you choose this dataset.

This dataset is called Census Income, and its purpose is to predict whether an individual's income exceeds $50,000 per year based on census data. I found this dataset on the UC Irvine Machine Learning Repository. The URL is https://archive.ics.uci.edu/dataset/20/census+income. I chose this dataset because I am interested in identifying good predictors of income and determining how accurate these predictors can be.

______________
# Data Exploration
Import your dataset into your .ipynb, create dataframes, and explore your data.  

Include:

* Summary statistics means, medians, quartiles,
* Missing value information
* Any other relevant information about the dataset.  



In [1]:
import pandas as pd
import numpy as np

#import data and add column names since there were no column names
df = pd.read_csv("https://raw.githubusercontent.com/yli1048/yli1048/refs/heads/602/adult.csv", names = ["age", "workclass", "fnlwgt", "education", "education_num", "marital_status", "occupation", "relationship", "race", "sex", "capital_gain", "capital_loss", "hours_per_week", "native_country", "income"], header = 0)

In [2]:
#check dataframe
print(df.head(3))

   age          workclass  fnlwgt   education  education_num  \
0   50   Self-emp-not-inc   83311   Bachelors             13   
1   38            Private  215646     HS-grad              9   
2   53            Private  234721        11th              7   

        marital_status          occupation    relationship    race    sex  \
0   Married-civ-spouse     Exec-managerial         Husband   White   Male   
1             Divorced   Handlers-cleaners   Not-in-family   White   Male   
2   Married-civ-spouse   Handlers-cleaners         Husband   Black   Male   

   capital_gain  capital_loss  hours_per_week  native_country  income  
0             0             0              13   United-States   <=50K  
1             0             0              40   United-States   <=50K  
2             0             0              40   United-States   <=50K  


In [3]:
#check for NA values in the dataframe
x = df.isnull().sum()
print(x)

age               0
workclass         0
fnlwgt            0
education         0
education_num     0
marital_status    0
occupation        0
relationship      0
race              0
sex               0
capital_gain      0
capital_loss      0
hours_per_week    0
native_country    0
income            0
dtype: int64


In [4]:
#find dimension
df.shape

(32560, 15)

In [5]:
#Find the mean for all numeric columns in dataframe
print(df.mean(numeric_only=True))

age                   38.581634
fnlwgt            189781.814373
education_num         10.080590
capital_gain        1077.615172
capital_loss          87.306511
hours_per_week        40.437469
dtype: float64


In [6]:
#Find the quartiles for all numeric columns in dataframe
print(df.quantile(q=[0, 0.25, 0.5, 0.75, 1], axis=0, numeric_only=True))

       age     fnlwgt  education_num  capital_gain  capital_loss  \
0.00  17.0    12285.0            1.0           0.0           0.0   
0.25  28.0   117831.5            9.0           0.0           0.0   
0.50  37.0   178363.0           10.0           0.0           0.0   
0.75  48.0   237054.5           12.0           0.0           0.0   
1.00  90.0  1484705.0           16.0       99999.0        4356.0   

      hours_per_week  
0.00             1.0  
0.25            40.0  
0.50            40.0  
0.75            45.0  
1.00            99.0  


This dataset does not contain any missing values. However, it includes values represented as "?".

This dataset was derived from the 1994 Census Bureau database by Ronny Kohavi and Barry Becker. Consequently, the information contained within this dataset may not serve as reliable predictors for contemporary societal trends or occupational classifications.

# Data Wrangling
Create a subset of your original data and perform the following.  

1. Modify multiple column names.

2. Look at the structure of your data – are any variables improperly coded? Such as strings or characters? Convert to correct structure if needed.

3. Fix missing and invalid values in data.

4. Create new columns based on existing columns or calculations.

5. Drop column(s) from your dataset.

6. Drop a row(s) from your dataset.

7. Sort your data based on multiple variables.

8. Filter your data based on some condition.

9. Convert all the string values to upper or lower cases in one column.

10. Check whether numeric values are present in a given column of your dataframe.

11. Group your dataset by one column, and get the mean, min, and max values by group.
  * Groupby()
  * agg() or .apply()

12. Group your dataset by two columns and then sort the aggregated results within the groups.

**You are free (and should) to add on to these questions.  Please clearly indicate in your assignment your answers to these questions.**

In [7]:
#create a subset for all the census respondents who were born in the United States
us = df[df["native_country"] == " United-States"].copy()

In [8]:
#modify multiple column names
us.columns = ["age", "workclass", "final_weight", "education", "education_level", "marital_status", "occupation", "relationship", "race", "gender", "capital_gain", "capital_loss", "hours_per_week", "native_country", "income"]
print(us.head(3))

   age          workclass  final_weight   education  education_level  \
0   50   Self-emp-not-inc         83311   Bachelors               13   
1   38            Private        215646     HS-grad                9   
2   53            Private        234721        11th                7   

        marital_status          occupation    relationship    race gender  \
0   Married-civ-spouse     Exec-managerial         Husband   White   Male   
1             Divorced   Handlers-cleaners   Not-in-family   White   Male   
2   Married-civ-spouse   Handlers-cleaners         Husband   Black   Male   

   capital_gain  capital_loss  hours_per_week  native_country  income  
0             0             0              13   United-States   <=50K  
1             0             0              40   United-States   <=50K  
2             0             0              40   United-States   <=50K  


All variables were correctly coded, as they were properly cleaned during the extraction process.

In [9]:
#check dimension
us.shape

(29169, 15)

In [10]:
#replace all ? to NA
us = us.replace(to_replace=" ?", value="NA")

In [11]:
#remove all "NA" values in the workclass column
value = ["NA"]
us = us[us.workclass.isin(value) == False]

In [12]:
#remove all "NA" values in the occupation column
us = us[us.occupation.isin(value) == False]

In [13]:
#check new dimension
print(us.shape)

(27503, 15)


In [14]:
#create new columns based on existing columns and calculation
us["capital_income"] = (us["capital_gain"] - us["capital_loss"])
print(us.head(3))

   age          workclass  final_weight   education  education_level  \
0   50   Self-emp-not-inc         83311   Bachelors               13   
1   38            Private        215646     HS-grad                9   
2   53            Private        234721        11th                7   

        marital_status          occupation    relationship    race gender  \
0   Married-civ-spouse     Exec-managerial         Husband   White   Male   
1             Divorced   Handlers-cleaners   Not-in-family   White   Male   
2   Married-civ-spouse   Handlers-cleaners         Husband   Black   Male   

   capital_gain  capital_loss  hours_per_week  native_country  income  \
0             0             0              13   United-States   <=50K   
1             0             0              40   United-States   <=50K   
2             0             0              40   United-States   <=50K   

   capital_income  
0               0  
1               0  
2               0  


In [15]:
#drop columns "capital_gain" and "capital_loss" since we already calculated the total
us = us.drop(["capital_gain", "capital_loss"], axis=1)
print(us.head(3))

   age          workclass  final_weight   education  education_level  \
0   50   Self-emp-not-inc         83311   Bachelors               13   
1   38            Private        215646     HS-grad                9   
2   53            Private        234721        11th                7   

        marital_status          occupation    relationship    race gender  \
0   Married-civ-spouse     Exec-managerial         Husband   White   Male   
1             Divorced   Handlers-cleaners   Not-in-family   White   Male   
2   Married-civ-spouse   Handlers-cleaners         Husband   Black   Male   

   hours_per_week  native_country  income  capital_income  
0              13   United-States   <=50K               0  
1              40   United-States   <=50K               0  
2              40   United-States   <=50K               0  


In [16]:
#drop the first two rows by index
us = us.drop([0, 1])
print(us.head(3))

   age          workclass  final_weight education  education_level  \
2   53            Private        234721      11th                7   
4   37            Private        284582   Masters               14   
6   52   Self-emp-not-inc        209642   HS-grad                9   

        marital_status          occupation relationship    race   gender  \
2   Married-civ-spouse   Handlers-cleaners      Husband   Black     Male   
4   Married-civ-spouse     Exec-managerial         Wife   White   Female   
6   Married-civ-spouse     Exec-managerial      Husband   White     Male   

   hours_per_week  native_country  income  capital_income  
2              40   United-States   <=50K               0  
4              40   United-States   <=50K               0  
6              45   United-States    >50K               0  


In [17]:
#sort the dataframe based on multiple variables
us = us.sort_values(by=["age", "education_level"], ignore_index=True)
print(us.head(3))

   age workclass  final_weight education  education_level  marital_status  \
0   17   Private        168807   7th-8th                4   Never-married   
1   17   Private         46402   7th-8th                4   Never-married   
2   17   Private        191260       9th                5   Never-married   

       occupation    relationship    race gender  hours_per_week  \
0    Craft-repair   Not-in-family   White   Male              45   
1           Sales       Own-child   White   Male               8   
2   Other-service       Own-child   White   Male              24   

   native_country  income  capital_income  
0   United-States   <=50K               0  
1   United-States   <=50K               0  
2   United-States   <=50K            1055  


In [18]:
#filter the dataframe based on having child condition
us_kids = us[us["relationship"] == " Own-child"].copy()
print(us_kids.head(3))

   age   workclass  final_weight education  education_level  marital_status  \
1   17     Private         46402   7th-8th                4   Never-married   
2   17     Private        191260       9th                5   Never-married   
4   17   Local-gov         32124       9th                5   Never-married   

       occupation relationship    race gender  hours_per_week  native_country  \
1           Sales    Own-child   White   Male               8   United-States   
2   Other-service    Own-child   White   Male              24   United-States   
4   Other-service    Own-child   Black   Male               9   United-States   

   income  capital_income  
1   <=50K               0  
2   <=50K            1055  
4   <=50K               0  


In [19]:
#convert all the string values to lower cases in column "workclass"
us_kids["workclass"] = us_kids["workclass"].str.lower()
print(us_kids.head(3))

   age   workclass  final_weight education  education_level  marital_status  \
1   17     private         46402   7th-8th                4   Never-married   
2   17     private        191260       9th                5   Never-married   
4   17   local-gov         32124       9th                5   Never-married   

       occupation relationship    race gender  hours_per_week  native_country  \
1           Sales    Own-child   White   Male               8   United-States   
2   Other-service    Own-child   White   Male              24   United-States   
4   Other-service    Own-child   Black   Male               9   United-States   

   income  capital_income  
1   <=50K               0  
2   <=50K            1055  
4   <=50K               0  


In [20]:
#check whether numeric values are present in a given column of the dataframe
print(us_kids.dtypes)

age                 int64
workclass          object
final_weight        int64
education          object
education_level     int64
marital_status     object
occupation         object
relationship       object
race               object
gender             object
hours_per_week      int64
native_country     object
income             object
capital_income      int64
dtype: object


In [21]:
#group the dataset by "marital_status", and get the mean, min, and max values by group
result = us_kids.groupby("marital_status").agg({"capital_income":["mean", "min", "max"]})
print("Mean, min, and max values of capital income grouped by marital status")
print(result)

Mean, min, and max values of capital income grouped by marital status
                      capital_income             
                                mean   min    max
marital_status                                   
Divorced                  176.180272 -1876  14344
Married-AF-spouse           0.000000     0      0
Married-civ-spouse        525.109589 -2129   7688
Married-spouse-absent     -55.241379 -1602      0
Never-married             114.540438 -2339  99999
Separated                 -31.938272 -3900   2597
Widowed                     0.000000     0      0


In [22]:
#group the dataset by two columns("race" and "gender") and then sort the aggregated results within the groups
result_2 = us_kids.groupby(["race", "gender"]).agg({"hours_per_week":["mean", "min", "max"]})
print("Mean, min, and max values of hours of work per week grouped by race and gender")
print(result_2)

Mean, min, and max values of hours of work per week grouped by race and gender
                          hours_per_week        
                                    mean min max
race               gender                       
Amer-Indian-Eskimo Female      27.750000   4  40
                   Male        37.423077  15  84
Asian-Pac-Islander Female      33.972222  12  50
                   Male        36.368421  16  60
Black              Female      33.648402   2  75
                   Male        34.822314   1  80
Other              Female      32.700000  15  65
                   Male        35.125000  12  52
White              Female      30.250160   2  99
                   Male        35.121302   2  99


In [23]:
#create a new subset for all people whose income is greater than 50k
df_50k = df[df["income"] == " >50K"].copy()
print(df_50k.head(3))

   age          workclass  fnlwgt   education  education_num  \
6   52   Self-emp-not-inc  209642     HS-grad              9   
7   31            Private   45781     Masters             14   
8   42            Private  159449   Bachelors             13   

        marital_status        occupation    relationship    race      sex  \
6   Married-civ-spouse   Exec-managerial         Husband   White     Male   
7        Never-married    Prof-specialty   Not-in-family   White   Female   
8   Married-civ-spouse   Exec-managerial         Husband   White     Male   

   capital_gain  capital_loss  hours_per_week  native_country income  
6             0             0              45   United-States   >50K  
7         14084             0              50   United-States   >50K  
8          5178             0              40   United-States   >50K  


In [24]:
#check for dimension
df_50k.shape

(7841, 15)

In [31]:
#replace all ? to NA
df_50k = df_50k.replace(to_replace=" ?", value="NA")

In [32]:
#remove NA values
df_50k = df_50k[df_50k.workclass.isin(value) == False]
df_50k = df_50k[df_50k.occupation.isin(value) == False]

In [33]:
#check for new dimension
df_50k.shape

(7650, 15)

In [34]:
#find the counts for the races of people whose income is greater than 50k
race_50k = df_50k["race"].value_counts().rename_axis("race").reset_index(name="counts")
print(race_50k)

                  race  counts
0                White    6944
1                Black     378
2   Asian-Pac-Islander     271
3   Amer-Indian-Eskimo      34
4                Other      23


In [35]:
#calculate percentage for each race
total = race_50k["counts"].sum()
race_50k["percentage"] = race_50k["counts"] / total * 100
print(race_50k)

                  race  counts  percentage
0                White    6944   90.771242
1                Black     378    4.941176
2   Asian-Pac-Islander     271    3.542484
3   Amer-Indian-Eskimo      34    0.444444
4                Other      23    0.300654


In [37]:
#find the counts for the occuppations of people whose income is greater than 50k
job_50k = df_50k["occupation"].value_counts().rename_axis("occupation").reset_index(name="counts")
print(job_50k)

            occupation  counts
0      Exec-managerial    1968
1       Prof-specialty    1859
2                Sales     983
3         Craft-repair     929
4         Adm-clerical     507
5     Transport-moving     320
6         Tech-support     283
7    Machine-op-inspct     250
8      Protective-serv     211
9        Other-service     137
10     Farming-fishing     115
11   Handlers-cleaners      86
12        Armed-Forces       1
13     Priv-house-serv       1


In [38]:
#calculate percentage for each occupation
total_2 = job_50k["counts"].sum()
job_50k["percentage"] = job_50k["counts"] / total_2 * 100
print(job_50k)

            occupation  counts  percentage
0      Exec-managerial    1968   25.725490
1       Prof-specialty    1859   24.300654
2                Sales     983   12.849673
3         Craft-repair     929   12.143791
4         Adm-clerical     507    6.627451
5     Transport-moving     320    4.183007
6         Tech-support     283    3.699346
7    Machine-op-inspct     250    3.267974
8      Protective-serv     211    2.758170
9        Other-service     137    1.790850
10     Farming-fishing     115    1.503268
11   Handlers-cleaners      86    1.124183
12        Armed-Forces       1    0.013072
13     Priv-house-serv       1    0.013072


# Conclusions  

After exploring your dataset, provide a short summary of what you noticed from this dataset.  What would you explore further with more time?

Upon analyzing the dataset, I identified that a limited percentage of individuals earn more than 50,000 annually. Notably, 90% of this subgroup comprises white individuals, which can be attributed to their increased likelihood of attaining higher education and having access to diverse resources. Furthermore, 25% of those earning over 50,000 occupy positions as executives or managers, while 24% work as specialized professors. Given additional time, I intend to further investigate potential correlations between higher education and factors such as marital status and gender.